src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -48,392 +55,1156 @@ internal static class ThreadPoolGlobals
+
+                                // make the slot appear empty in the next generation
+                                Volatile.Write(ref slot.SequenceNumber, position + 1 + _slotsMask);
+                                return item;
+                            }
                         }
+                        else if (sequenceNumber - position < Full)
+                        {
+                            // The sequence number was less than what we needed, which means we cannot return current item.
+                            // Check if we have reached Enqueue and return null.
+                            // Otherwise spin.
+                            // NB: reading stale _frozenForEnqueues is fine - we would just spin once more
+                            var currentEnqueue = _queueEnds.Enqueue;
+                            if (currentEnqueue == position || (_frozenForEnqueues && currentEnqueue == position + FreezeOffset))
+                            {
+                                return null;
+                            }
+                        }
+
+                        // Lost a race. Spin a bit, then try again.
+                        spinner.SpinOnce();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// The "local" flavor of the queue is similar to the "global", but also supports Pop and Remove operations.
+        /// 
+        /// - Pop is used to implement Busy-Leaves scheduling strategy.
+        /// - Remove is used when the caller finds it benefitial to execute a workitem "inline" after it has been scheduled. 
+        ///   (such as waiting on a task completion).
+        /// 
+        /// We create multiple local queues and softly affinitize them with CPU cores.
+        /// </summary>
+        [DebuggerDisplay("Count = {Count}")]
+        internal sealed class LocalQueue : WorkQueueBase
+        {
+            /// <summary>The current enqueue segment.</summary>
+            internal LocalQueueSegment _enqSegment;
+            /// <summary>The current dequeue segment.</summary>
+            internal LocalQueueSegment _deqSegment;
+
+            [StructLayout(LayoutKind.Explicit, Size = 2 * Internal.PaddingHelpers.CACHE_LINE_SIZE)] // padding before/after fields
+            private struct Rnd
+            {
+                [FieldOffset(1 * Internal.PaddingHelpers.CACHE_LINE_SIZE)] internal uint val;
+            }
+
+            // Very cheap random sequence generator. We keep one per-local queue.
+            private Rnd _rnd = new Rnd() { val = 6247 };
+
+            // We do not need a lot of randomness, I think even _rnd++ would be fairly good here. 
+            // Sequences attached to different queues go out of sync quickly and that could be sufficient.
+            // However this sequence is a bit more random at a very modest additional cost.
+            // http://www.drdobbs.com/tools/fast-high-quality-parallel-random-number            
```

Comment:
This link doesn't work anymore. I've found that permalink to that page is http://www.drdobbs.com/tools/fast-high-quality-parallel-random-number/231000484#

Reply:
The link used to work. I guess something changed on the other side. Will have to update the link.
 Thanks!!